### ISSUE TITLE
includegraphics filename autocompletion does not follow symlinks

### ISSUE THREAD

**Issue by michaelpacer** [SublimeText/LaTeXTools#792]
When you try to use `\includegraphics{` a generally useful popup menu arises that does fuzzy matching on what you type. 

However, if you are trying to jump into a symlinked directory, it does not seem to even notice that the directory is there. 

I'm not sure if its a separate issue but if you then proceed to type the name of the exact file path and hit enter(when the search field says nothing is there), it won't select what you have written and place it in the text. Worse, even if you press enter repeatedly, the popup dialog will not go away.  (I'll make this a separate issue just in case and will close it if its not).

**Comment by r-stein** [comment]
This should easiliy solved by adding `followlinks=True` [here](https://github.com/SublimeText/LaTeXTools/blob/v3.9.3/latex_input_completions.py#L155).  ([python documentation](https://docs.python.org/3.3/library/os.html#os.walk))
However I am worried this could result in cyclic paths/infinite recursion.

**Comment by ig0774** [comment]
We can always use `os.path.realpath()` and store the processed directories. Something like:

```python
handled_directories = set()
for dir_name, dirs, files in os.walk(path, followlinks=True):
    handled_directories.add(os.path.realpath(dir_name))
    files = [f for f in files if f[0] != '.' and file_match(f)]
    dirs[:] = [d for d in dirs if d[0] != '.' and
                   os.path.join(dir_name, d) != output_directory and
                   os.path.join(dir_name, d) != aux_directory and
                   os.path.realpath(os.path.join(dir_name, d)) not in handled_directories]
    ...
```

The trade-off is a few `os.stat` calls in normal cases and potentially exhausting memory in truly pathological cases (i.e. where a subfolder symlinks to root)...

**Comment by ig0774** [comment]
Fixed in 3.10.0.

**Comment by michaelpacer** [comment]
Thanks for the responsiveness — I can't figure out looking at your code how you avoided the root problem…is that just the `if d[0] != '.'` clause plus the `dir_match` function?

also is the pathological case just when its at the root of any tree or anywhere there my be a symlink cycle?

**Comment by ig0774** [comment]
As @r-stein mentioned, its the `followlinks` that actually gets `os.walk` to descend into symlink dirs. The code I added simply uses a set to track the non-symlinked path of the dirs we've already checked. The `dir_match` function then eliminates any subdirectories that are symlinks to a path we already checked. That takes care of the symlink cycle.

The pathological case I was thinking of is where a symlink links to a directory that causes the set we use for tracking to consume all available memory. It's exceedingly unlikely, it's just not impossible. A symlink to `/` (the root directory) as a subdirectory will cause `os.walk` to descend through every (readable) directory in the file system, which means we could potentially end up storing thousands of directory strings...
